Improve compress record

simon-ess opened the pull request:
This pull request accomplishes two main points:

1. Add testing to the compress record. This record had only very limited testing of two of the algorithms that are used. This pull request adds in more testing for the existing algorithms.
2. This also adds a new field (PBUF) which allows a user to select if partially filled buffers should be nevertheless calculated. For example, if you have a periodically scanned ai record that is being averaged, at the moment it waits until the entire buffer is full before it produces an average. With this change and with PBUF set to YES, then the average will be calculated based on the running tally.

See https://epics.anl.gov/tech-talk/2021/msg00239.php for some context.

Note that there are some other potential improvements to this record: for example, if you have a waveform record as input to a compress record, and you fill up the compress record and process it, then an update to the waveform record that does not fill it completely followed by another process of the compress record will yield no obvious clue that the compress record was not processed.

We could set the compress record in such a case to UDF or some other alarm state. That said, that feels out of scope for the suggested change.

simon-ess committed:
Initial test for averaging

simon-ess committed:
Added failing test for partially filled buffer average

simon-ess committed:
Add test path for single input data

simon-ess committed:
Single input data test passes

simon-ess committed:
Added test for low value

simon-ess committed:
Added failing test for partial ai average

simon-ess committed:
compress_scalar for average works correctly now

simon-ess committed:
Added test for array_average

exzombie commented on modules/database/test/std/rec/compressTest.c:
```diff
@@ -346,10 +348,292 @@ void testLIFOCirc(void)
+
+void testArrayAverage(void) {
+    DBADDR wfaddr, caddr;
+    
```
Extraneous whitespace. Very important ;)

exzombie commented on modules/database/src/std/rec/compressRecord.dbd.pod:
```diff
@@ -385,7 +385,15 @@ Scan forward link if necessary, set PACT FALSE, and return.
 		menu(compressALG)
 	}
- 	field(BALG,DBF_MENU) {
+	field(PBUF,DBF_MENU) {
+		prompt("Use Partial buffers")
+		promptgroup("30 - Action")
+		special(SPC_RESET)
+		interest(1)
+		menu(menuYesNo)
+		initial("NO")
+	}
```
The new behavior still needs to be documented, both in release notes and here. I know you know this, just adding a reminder so that we don't forget about it.

simon-ess replied:
I have updated the documentation, let me know if this seems sufficient.

exzombie commented on modules/database/test/std/rec/compressTest.c:
```diff
@@ -346,10 +348,292 @@ void testLIFOCirc(void)
+    va_end(args);
+
+    dbScanLock(addr->precord);
+    testOk1(dbPut(addr, DBR_DOUBLE, values, count)==0);
```
As I see it, `writeToWaveform` is a utility function, and it's only used to write to the waveform record. Thus, I don't think this `dbPut` needs to be checked because it can only fail if the test itself is broken, which has nothing to do with the code under test. Testing it only complicates counting your tests.

exzombie commented on modules/database/test/std/rec/compressTest.c:
```diff
@@ -346,10 +348,292 @@ void testLIFOCirc(void)
+}
+
+void
+testAIPartialAverage(void) {
```
I'd name the function `testAIAveragePartial` to maintain consistency with other tests.

exzombie commented on modules/database/test/std/rec/compressTest.c:
```diff
@@ -346,10 +348,292 @@ void testLIFOCirc(void)
+    int i;
+    DBADDR aiaddr, caddr;
+
+    testDiag("Test 'N to 1 Low Value'");
```
The string naming this test is from a previous test.

exzombie commented on modules/database/test/std/rec/compressTest.c:
```diff
@@ -346,10 +348,292 @@ void testLIFOCirc(void)
+
+    for (i = 0; i < 4; i++) {
+        dbScanLock(aiaddr.precord);
+        testOk1(dbPut(&aiaddr, DBR_DOUBLE, &data[i], 1) == 0);
```
See comment on `writeToWaveform`, same considerations apply.

exzombie commented on modules/database/test/std/rec/compressTest.c:
```diff
@@ -346,10 +348,292 @@ void testLIFOCirc(void)
+
+    testDiag("Test partial data with PBUF set to YES");
+
+    ((compressRecord *)caddr.precord)->pbuf = menuYesNoYES;
```
Record fields need to be accessed with the record locked.

exzombie commented:
I found no issues with the changes to the record.

As for tests, it's great that you added some, and I only have minor comments. As this is a strict improvement over the previous state, I think it could be merged as-is. That said, the tests could be improved further by making them less repetitive. We can discuss how it could be done if there is interest in doing so.

I agree that putting the record to UDF is out of scope. But for the record (har, har) and future discussion, I wonder how much of a compatibility issue such a change would introduce. On one hand, it would make things more "poper", but it's not worth introducing issues over that. On the other hand, I can't imagine a use case where compress record's STAT becoming UDF would cause any problems, apart from being indicated on an operator screen when it wasn't before. But I guess that just means I'm not imaginative enough :)

simon-ess committed:
General cleanup

simon-ess committed:
Fixing 'error C2057: expected constant expression' error

simon-ess committed:
Added one more put/process to go over the buffer length

Note that it is not really a circular buffer in this case, but
a full reset of the buffer to the beginning. This matches the
documentation, but it seems valuable to add an explicit test for
this case.

simon-ess committed:
Updated documentation

simon-ess commented:
I have updated the documentation, let me know if this seems sufficient.

exzombie commented:
As a user looking at the docs of the PBUF field, I'd intuitively expect it to also work with the `Average` algorithm, making it calculate a moving average. This didn't occur to me yesterday because the use case described on the mailing list calculates exactly that. But without that context, it feels like an omission. I'm not necessarily suggesting implementing a moving average, it feels a bit like scope creep. But perhaps it could be clearly stated in the docs that PBUF does not affect the `Average` algorithm and that one should instead combine a `Circular Buffer` with `N to 1 Average` to achieve the desired effect.

simon-ess commented:
@exzombie - I had thought the same thing as well. The "Average" algorithm is a bit strange to me in that I would have expected it to behave exactly as you said. However, experimentation and a closer reading showed a totally different behaviour...

Anyhow, I will update the documentation as you suggest. I think that is a good addition.

simon-ess committed:
Added commend about PBUF having no effect on 'Average' algorithm

anjohnson commented:
Group 2023-03-06: Approved subject to adding a Release Notes entry.

simon-ess committed:
Updated RELEASE_NOTES

anjohnson commented:
LGTM, thanks.

mdavidsaver commented:
Applied as of e5ad12e638629ad814c1eb158e742951c6347db1
